### DISCUSSION TITLE
Input expression conflict, timeouts, tap function.

### DISCUSSION THREAD

**Discussion by pavvd** [houmain/keymapper#92] [Q&A] [answered]
As my cfg expands I ran into problem, where after adding hokteys like Caps+Alt+[some_key], tap function of CapsLock breaks. No matter how I rearrange things, I can't make them both work.

--- part of cfg ---
```
(CapsLock AltL){J} >> Super{ArrowLeft}     #--
CapsLock {AltL{K}} >> Super{ArrowRight}    #-- recently added

CapsLock{!250ms} >> Escape                 #-- no longer works
AltL{!250ms} >>Backspace                   #-- this still works
CapsLock{Any} >> Control{Any}
CapsLock >> Control
Control >> CapsLock

MetaLeft >> AltLeft
AltLeft{Any} >> MetaLeft{Any}
AltLeft >> MetaLeft
```

**Comment by houmain** [comment]
Hi,
I played a little bit with your configuration. After moving the first two mappings down, I could trigger Escape and Control with the CapsLock key and also the Super{ArrowLeft}.
```
CapsLock{!250ms} >> Escape                 #-- no longer works
AltL{!250ms} >>Backspace                   #-- this still works

(CapsLock AltL){J} >> Super{ArrowLeft}     #--
CapsLock {AltL{K}} >> Super{ArrowRight}    #-- recently added

CapsLock{Any} >> Control{Any}
CapsLock >> Control
Control >> CapsLock

MetaLeft >> AltLeft
AltLeft{Any} >> MetaLeft{Any}
AltLeft >> MetaLeft
```

**Comment by pavvd** [comment]
Thank you for your reply. As I do more testing using above order, results get more bizarre on my end.
(CapsLock AltL){key} -- works only if I hold Alt 1st then Caps.
CapsLock{AtlL{key}} -- still doesn't work
Putting Alt before Caps (AltL CapsLock){key} -- makes no difference.

I'm running Mint 21.2 with Awesome WM and no other remapping tools. It's probably something with my setup.

**Reply by houmain** [reply]
Could you please create a minimum configuration with which one can reproduce the bizarre behaviour? Thanks!

**Comment by pavvd** [comment]
Tested with only those 4 lines in cfg:
```
CapsLock{!250ms} >> Escape

(CapsLock AltLeft){J} >> MetaLeft{ArrowLeft}   #-- 1

CapsLock{Any} >> Control{Any}
CapsLock >> Control
```
1 works only if Alt is held 1st then Caps and [key] or both mods pressed simultaneously.
AltLeft{CapsLock{K}} >> MetaLeft{ArrowLeft} -- this works exactly the same as 1.
CapsLock {AltLeft{K}} >> MetaLeft{ArrowLeft} -- this doesn't work at all.

Side note:  I don't run keymapper with -u option, just kill/reset everything with shell script.

**Reply by houmain** [reply, marked as answer]
I think I know what happens:
Holding `CapsLock` for a while presses `Control`, which is kept pressed until `CapsLock` is released.
Then pressing `AltLeft` and `J` triggers `MetaLeft{ArrowLeft}`.
So effectively `Control{MetaLeft{ArrowLeft}}` is pressed which means nothing to your WM.

Adding a `!Control` should make it work:
`(CapsLock AltLeft){J} >> !Control MetaLeft{ArrowLeft}`

**Reply by pavvd** [reply]
> Adding a !Control should make it work:
(CapsLock AltLeft){J} >> !Control MetaLeft{ArrowLeft}

Works like a charm! I would never figure it out myself. Thank you very much.
